lib/large_object_store.rb

```diff
@@ -53,6 +55,11 @@ def read(key)
         return nil if slices.compact.size < pages
         slices.join("")
       end
+
+      if data.getbyte(0) == 0x78 && [0x01,0x9C,0xDA].include?(data.getbyte(1))
```

Comment:
Rails serializes a class called `Entry` which has a `@compressed` member indicating whether `entry.value` is compressed:

https://github.com/rails/rails/blob/6c61ae816ad7326e510a8f019cadc6db4cc9dae0/activesupport/lib/active_support/cache.rb#L601-L617

https://github.com/rails/rails/blob/6c61ae816ad7326e510a8f019cadc6db4cc9dae0/activesupport/lib/active_support/cache/mem_cache_store.rb#L176-L183

One of the consequences of the Rails approach is they end up double-serializing, or at least copying the memory around twice, since `Marshal.dump` is called prior to compression and then called again on the compressed value.

We could take the same approach of serializing a class `Entry` that wraps the real value; it'd be a little safer than checking for zlib magic at the beginning. But it'd make this change a lot more complicated.